include/boost/url/bnf/parse.hpp

```diff
@@ -77,214 +96,182 @@ parse(
+   @param it An iterator to the first element in the range.
+   At the end of this function, `it` points to one past the last
    element parsed in the range.
 
-   @param[in] end An iterator to one past the last element in the range
+   @param end An iterator to one past the last element in the range
 
-   @param[out] ec The error code this function should set if it fails
+   @param ec The error code this function should set `it` it fails
 
-   @param[out] args BNF elements to be parsed
+   @param args BNF elements to be parsed
 
 */
+template<
 #ifdef BOOST_URL_DOCS
-template<class... Args>
-bool parse( char const*& it, char const* end, Args&&... args);
+    class... Args
 #else
-template<
     class T0,
     class T1,
-    class... Tn>
+    class... Tn
+#endif
+>
 bool
 parse(
     char const*& it,
     char const* end,
     error_code& ec,
+#ifdef BOOST_URL_DOCS
+    Args&&... args
+#else
     T0&& t0,
     T1&& t1,
-    Tn&&... tn);
+    Tn&&... tn
 #endif
+);
 
-/** @brief Parse a complete string
+/** Parse a complete string as a sequence of BNF elements
 
-   This function parses a complete string @ref s into the Backus–Naur form (BNF)
-   elements [@ref t0, @ref t1, ...]. If the string is not completely consumed,
-   @ref ec is set to @ref error::leftover_input.
+   This function parses a complete string `s` into the Backus–Naur form (BNF)
+   elements [`arg_0`, `arg_1`, ...] in `args`.
 
-   The function returns `true` and resets @ref ec when the input is
-   parsed successfully into the elements [@ref t0, @ref t1, ...].
+   @par Example
 
-   If the input cannot be parsed into the BNF elements, the function returns
-   `false` and sets the error number in @ref ec to the appropriate error
-   number defined in @ref boost::urls::error.
+   @code
+   if (bnf::parse(str, ec, bnf_element_1, bnf_element_2, bnf_element_3)) {
+     std::cout << "String " << str << " parsed successfully\n";
+   }
+   @endcode
 
    @par Exception Safety
-     Calls to `parse` overloads may throw, although they should
-     prefer setting @ref ec instead.
 
-   @see @ref boost::urls::error
+     Internal calls to `parse` overloads may throw, although they should
+     prefer setting `ec` instead.
 
-   @return `true` if the string matches the BNF elements successfully and all
-   chars are consumed, `false` otherwise
+   @see
 
-   @param[in] s The input string this function should parse
+   [Backus–Naur form](https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form),
 
-   @param[out] ec The error code this function should set if it fails
+   @ref boost::urls::error
 
-   @param[out] t0 T0 1st BNF element
+   @return `true` if the string matches the BNF elements successfully and all
+   chars are consumed, `false` otherwise.
+
+   @param s The input string this function should parse
 
-   @param[out] t1 T1 2nd BNF element
+   @param ec The error code this function should set if it fails. If the string is not completely consumed,
```

Comment:
Parameter descriptions need to be short. `error::leftover_input` should be a `@ref`. I plan on changing the name. And it will be in the bnf namespace enum not the urls one.

Reply:
Oh... I did try that, but I got:

```
/home/alandefreitas/Documents/Code/C++/boost/libs/url/include/boost/url/bnf/parse.hpp:190: warning: unable to resolve reference to `error::leftover_input' for \ref command
```

It was because the name has already been changed between the first version of the javadocs and now. It works now! :D